src/nodes/statements.ts

```diff
@@ -71,17 +71,48 @@ export class CEmptyStatement {
 }
 
 @CodeTemplate(`
-{destructors}
-return {expression};
+{#if returnTypeAndVar}
+    {
+        {returnTypeAndVar} = {expression};
+        {destructors}
+        return {returnTemp};
+    }
+{#else}
+    {destructors}
+    return {expression};
+{/if}
 `, ts.SyntaxKind.ReturnStatement)
 export class CReturnStatement extends CTemplateBase {
     public expression: CExpression;
     public destructors: CVariableDestructors;
     public retVarName: string = null;
+    public returnTypeAndVar: string = null;
+    public returnTemp: string = null;
     public closureParams: { name: string, value: CExpression }[] = [];
+    doCheckVarNeeded(scope: IScope, node: ts.ReturnStatement): boolean {
+        let s = scope.root.memoryManager.getDestructorsForScope(node);
+        let result = false;
+        for (let e of s) {
+            result = result || new RegExp('\\W' + e.varName + '\\W').test(' ' + node.getFullText() + ' ');
+        }
+        return result;
+    }
     constructor(scope: IScope, node: ts.ReturnStatement) {
         super();
+        this.returnTemp = scope.root.symbolsHelper.addTemp(node, 'returnVal');
+        let returnType = scope.root.typeHelper.getCType(node.expression);
+        if (this.doCheckVarNeeded(scope, node)) {
+            // todo: make this less hackyy
+            let fakeVar = new CVariable(scope, '__fake', returnType, { removeStorageSpecifier: true, arraysToPointers: true });;
+            let fakeVarType = fakeVar.resolve().slice(0, -6).trim();
+            this.returnTypeAndVar = fakeVarType;
+            if (this.returnTypeAndVar.indexOf('{var}') == -1) {
+                this.returnTypeAndVar += ' {var}';
+            }
+            this.returnTypeAndVar = this.returnTypeAndVar.replace('{var}', this.returnTemp);
```

Comment:
this is not only too hacky, it is also not a proper C89 as it produces a variable declaration that is not in the beginning of the block:
```
mem/mem7.c: In function ‘return_from_obj’:
mem/mem7.c:23:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   23 |     struct return_from_obj_t * returnVal = obj->key;
      |     ^~~~~~
```

it's very easy to fix though, you just need to do smth like this:
```
if (tempVarIsNeeded) {
    this.retVarName = scope.root.symbolsHelper.addTemp(node, 'returnVal');
    let returnType = scope.root.typeHelper.getCType(node.expression);
    scope.variables.push(new CVariable(scope, this.retVarName, returnType, { removeStorageSpecifier: true }))
}
```

pushing the variable into the scope variables will automatically add it's declaration into the right place